pkg/server/handler.go

```diff
@@ -305,6 +306,44 @@ func WithImpersonationGatekeeper(handler http.Handler) http.Handler {
 	})
 }
 
+// WithScoping scopes the request to the cluster it is intended for.
+func WithScoping(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		user, exists := request.UserFrom(req.Context())
+		if !exists {
+			responsewriters.ErrorNegotiated(
+				apierrors.NewForbidden(schema.GroupResource{}, "", fmt.Errorf("impersonation is invalid for the requestor")),
+				errorCodecs, schema.GroupVersion{}, w, req)
+			return
+		}
+
+		cluster := request.ClusterFrom(req.Context())
+		if cluster == nil {
+			responsewriters.ErrorNegotiated(
+				apierrors.NewForbidden(schema.GroupResource{}, "", fmt.Errorf("no cluster found in the context")),
+				errorCodecs, schema.GroupVersion{}, w, req)
+			return
+		}
+
+		extra := user.GetExtra()
+		if extra == nil {
+			extra = map[string][]string{}
+		}
+		extra["authentication.kcp.io/scopes"] = append(extra["authentication.kcp.io/scopes"], fmt.Sprintf("cluster:%s", cluster.Name))
```

Comment:
this is too much. We only have to do this if impersonation happened.

Reply:
Yes, was thinking that need to set some metadata for this and check it here.